021-presence-service.md

```diff
@@ -0,0 +1,163 @@
+
+## Resources
+<!-- List all the resources used for writing this RFC -->
+
+## Implementation
+
+<!-- Write an overview to explain the suggested implementation -->
+
+### API Changes
+<!-- Do we need new API endpoints? List and describe them and their API signatures -->
+
+```POST /v1/presences```
+
+Create a presence. Set the user presence status, expiry (minutes/date - TBD), and permissions for who has access to see the current presence. This will imitate the equivalent action in the GraphQL and Realtime APIs.
+
+```GET /v1/presences```
+
+List all presences, on client side this will show only presneces I have permission to view. On server it will show all presences. This endpoint will support queries and pagination like any other list endpoint on Appwrite.
+
+```GET /v1/presences/:id```
+
+Get presence by ID.
+
+```PUT /v1/presences/:id```
+
+Update presnece by ID.
+
+```DELETE /v1/presences/:id```
+
+Delete presence by ID.
+
+### Events
+
+The new endpoints will trigger standard events that will be used for listening to Realtime API changes or to execute functions and webhooks.
+
+### Console updates
+
+- We should implement presence in the console itself and make current org members online or not or last online
+- We will add a new section under auth called `Presences`, this section will show a dashboard with usage showing how many people were online in a specific time range. Below that, we will show avatars of all currently online users or users who just recently went offline. 
+
+###  Workers / Commands
+<!-- Do we need new workers or commands for this feature? List and describe them and their API signatures -->
+
+- The maintenance worker will need to clean all present documents that have expired. We should have a specific grace period that will allow us also to show people who just recently went offline. So if the presence is valid for 5 minutes, we will delete it after 5+ X minutes. X will be decided later. 
+- Usage service should count how many people were online at a specific time. 
+
+### Enable / Disable
+
+As with every Appwrite service, we should provide an option in the console to enable or disable it as it may not be needed for every use case.
+
+###  Supporting Libraries
+<!-- Do we need new libraries for this feature? Mention which, define the file structure and different interfaces -->
+
+Utopia/databases will need to expose Group by functionality so we can fetch only a unique list of presences per user as a user is allowed to report it multiple times to multiple groups. The most recent presence will take priority in case multiple presences allow a different user to view them. 
+
+### Data Structures
+<!-- Do we need new data structures for this feature? Describe and explain the new collections and attributes -->
+
+We need to add a new collection to store presences.
+
+- _id
+- _uid
+- _createdAt
+- _updatedAt
+- userInternalId
+- userId
+- expiry (DateTime)
+- status (string)
```

Comment:
I think this should be JSON, since often apps need more than just a message. GitHub lets you set emoji. Discord lets you also set game you currently play. Instagram lets you ass music to your note.